### PR TITLE
acgan: don't train discriminator to produce class labels for generated images

### PR DESCRIPTION
**Pull request by ozabluda**
I think it's almost completely conceptually wrong to train discriminator to produce class labels for generated images (except as a way to handicap generator). The biggest argument against merging this PR IMHO is that it deviates from the referenced paper.

The arguments for it:
1. Conceptually more correct. The rest follows because of this (I think).
2. Generated images are more diverse (most easily seen in 7), diversity is more robust epoch-to-epoch
3. Training is much more stable. Before, if generator falls behind even a bit, it's game over - all black images. Now generator easily recovers, even over long periods, even after hugely falling behind.
4. soft labels (handicapping discriminator) could be made much harder (works all the way to 0.90+)

Epoch 34 =========================== Epoch 36 (compare to https://github.com/fchollet/keras/pull/8383)
plot_epoch_034_generated plot_epoch_036_generated

### CODE REVIEW

**Comment by ozabluda** [comment]
Another approach is to introduce 11th class "generated" and train discriminator to output it on generated images. I tried, it also works, doesn't require soft labels, deviates from the referenced paper less, but the generated images are not as diverse.

It's interesting to notice that with all approaches, some info about real/fake is leaking into the auxiliary classifier. For example, with PR as it currently stands, auxiliary classifier loss is much larger on generated images (0.1) than on real images (0.01). I think this is a deficiency in generator training and the root cause for it is that generator aux loss (0.1) is still much smaller then generation loss (0.69 in Nash equilibrium). That could be relatively easily fixed, but would make the example more complicated without sufficient justification.

**Comment by fchollet** [comment]
@lukedeo, as the original author of this example, would you like to review this PR?

**Comment by ozabluda** [comment]
The following images were generated from linear interpolation of latent vector between top and bottom rows (both random) on epochs `plot_epoch_0{31,32,33,34,35,41,43,44,47,53}_generated.png` (in a setup of maybe subsequent PR, different from in this PR, which doesn't matter for the conclusion). This shows the usual (that the generation is meaningful - not just memorized, and latent space has structure), and doesn't show what it usually doesn't.

plot_epoch_031_generated  plot_epoch_032_generated
plot_epoch_033_generated  plot_epoch_035_generated
plot_epoch_041_generated  plot_epoch_043_generated
plot_epoch_044_generated  plot_epoch_047_generated
plot_epoch_053_generated  plot_epoch_061_generated
plot_epoch_068_generated

**Comment by lukedeo** [comment]
@fchollet will take a look later today

**Comment by ozabluda** [comment]
[OpenReview | Conditional Image Synthesis With Auxiliary Classifier GANs](https://openreview.net/forum?id=rJXTf9Bxg):

ICLR 2017 conference AnonReviewer1: 
>(2) In section 3 the AC-GAN classification objective (omitting expectation for brevity) is given as L_S = log P(C=c|X_real) + log P(C=c|X_fake) and you say that both the discriminator and generator are trained to maximize this quantity. Obviously the generator would want to maximize  log P(C=c|X_fake) for its given conditioning class c. But can you explain why you would want the discriminator to also maximize the classification accuracy of generated samples? Why not do something similar to the CatGAN paper [3] and train the discriminator to be as uncertain as possible about the generated examples. Seems counterintuitive to me to have both the generator and discriminator trying to optimize the same classification objective rather than not be adversarial wrt to this loss as well as the real/fake loss. 

Response to Reviewer 1:
>The short answer is that it makes the results better. Our suspicion is that this is due to a sort of feedback loop: more training data for the discriminator makes the discriminator better, which is then reflected in the generator.

I think this clearly removes my own biggest argument against merging this PR (see PR description).

**Comment by ozabluda** [comment]
Merged into https://github.com/fchollet/keras/pull/8482
